pkg/policy/selectorcache.go

```diff
@@ -695,132 +248,103 @@ func (l *labelIdentitySelector) fetchIdentityMappings() []identity.NumericIdenti
+type UpdateResult uint32
+
+const (
+	// UpdateResultUpdatePolicyMaps indicates the caller should call EndpointManager.UpdatePolicyMaps()
+	UpdateResultUpdatePolicyMaps = UpdateResult(1) << iota
+
+	// UpdateResultIdentitiesNeeded indicates the caller should wait for a complete ipcace label injection,
+	// as the selector is missing identities
+	UpdateResultIdentitiesNeeded
+
+	UpdateResultUnchanged = UpdateResult(0)
+)
+
 // UpdateFQDNSelector updates the mapping of fqdnKey (the FQDNSelector from a
-// policy rule as a string) to to the provided list of identities. If the contents
-// of the cachedSelections differ from those in the identities slice, all users
-// are notified asynchronously. Caller should Wait() on the returned
-// sync.WaitGroup before triggering any policy updates. Policy updates may need
-// Endpoint locks, so this Wait() can deadlock if the caller is holding any
-// endpoint locks.
-func (sc *SelectorCache) UpdateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity, wg *sync.WaitGroup) {
+// policy rule as a string) to to the provided list of IPs. If the contents
+// of the cachedSelections changes, all users are notified asynchronously.
+// Caller should Wait() on the returned sync.WaitGroup before triggering
+// any policy updates. Policy updates may need Endpoint locks, so this
+// Wait() can deadlock if the caller is holding any endpoint locks.
+//
+// Returns an UpdateResult that indicates if the caller should call UpdatePolicyMaps() and/or
+// wait for ipcache UpsertMetadata() to finish.
+func (sc *SelectorCache) UpdateFQDNSelector(fqdnSelec api.FQDNSelector, ips []netip.Addr, wg *sync.WaitGroup) UpdateResult {
 	sc.mutex.Lock()
-	identitiesToRelease := sc.updateFQDNSelector(fqdnSelec, identities, wg)
-	sc.mutex.Unlock()
-	sc.releaseIdentityMappings(identitiesToRelease)
+	defer sc.mutex.Unlock()
+	return sc.updateFQDNSelector(fqdnSelec, ips, wg)
 }
 
-func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identities []identity.NumericIdentity, wg *sync.WaitGroup) (identitiesToRelease []identity.NumericIdentity) {
-	fqdnKey := fqdnSelec.String()
-
-	var fqdnSel *fqdnSelector
+func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, ips []netip.Addr, wg *sync.WaitGroup) UpdateResult {
+	key := fqdnSelec.String()
 
-	selector, exists := sc.selectors[fqdnKey]
-	if !exists || selector == nil {
-		fqdnSel = &fqdnSelector{
-			selectorManager: selectorManager{
-				key:              fqdnKey,
-				users:            make(map[CachedSelectionUser]struct{}),
-				cachedSelections: make(map[identity.NumericIdentity]struct{}),
-			},
-			selector: fqdnSelec,
-		}
-		sc.selectors[fqdnKey] = fqdnSel
-	} else {
-		fqdnSel = selector.(*fqdnSelector)
+	idSelector, exists := sc.selectors[key]
+	if !exists || idSelector == nil {
+		log.WithField(logfields.Selector, fqdnSelec.String()).Error("UpdateFQDNSelector of selector not registered in SelectorCache!")
```

Comment:
Did something functionally change about the control flow such that this path no longer initializes new selectors? Or was the creation of a selector from this path previously just dead code that was never used?

If we have good paths to exercise this code then I'm not against this, but I'm not sure whether this path is being fully reasoned through. For instance, if the selector is not cached then why would this function continue and will that have unforeseen effects?

Reply:
Ah, I see that in the final patch this case should probably return due to the failure to cast to an `fqdnSelector`, so probably this is now fine... though I still don't fully understand why the previous code would create a new FQDN selector here.

Reply:
This is the same situation as https://github.com/cilium/cilium/pull/29036#discussion_r1406029335 -- because we had to unlock the selector cache during AddFQDNSelector, it was possible for updates to be received before the selector add was complete. This is no longer possible.